core/src/main/java/com/datastrato/gravitino/authorization/UserGroupManager.java

```diff
@@ -190,35 +170,78 @@ public boolean removeGroup(String metalake, String group) {
+  Group getGroup(String metalake, String group, Instant latestDropRoleTIme) {
     try {
       AuthorizationUtils.checkMetalakeExists(store, metalake);
-      return store.get(ofGroup(metalake, group), Entity.EntityType.GROUP, GroupEntity.class);
+
+      GroupEntity entity =
+          store.get(
+              NameIdentifierUtils.ofGroup(metalake, group),
+              Entity.EntityType.GROUP,
+              GroupEntity.class);
+
+      List<String> roles =
+          getValidRoles(
+              metalake,
+              latestDropRoleTIme,
+              entity.roles(),
+              entity.roleIds(),
+              entity.latestGcRolesTime());
+
+      return GroupEntity.builder()
+          .withId(entity.id())
+          .withName(entity.name())
+          .withAuditInfo(entity.auditInfo())
+          .withRoleNames(roles)
+          .build();
     } catch (NoSuchEntityException e) {
       LOG.warn("Group {} does not exist in the metalake {}", group, metalake, e);
-      throw new NoSuchGroupException(GROUP_DOES_NOT_EXIST_MSG, group, metalake);
+      throw new NoSuchGroupException(AuthorizationUtils.GROUP_DOES_NOT_EXIST_MSG, group, metalake);
     } catch (IOException ioe) {
       LOG.error("Getting group {} failed due to storage issues", group, ioe);
       throw new RuntimeException(ioe);
     }
   }
 
-  private NameIdentifier ofUser(String metalake, String user) {
-    return NameIdentifier.of(
-        metalake, Entity.SYSTEM_CATALOG_RESERVED_NAME, Entity.USER_SCHEMA_NAME, user);
-  }
+  private List<String> getValidRoles(
+      String metalake,
+      Instant latestDropRoleTime,
+      List<String> roleNames,
+      List<Long> roleIds,
+      Instant latestGcRoleTime) {
+    if (roleNames == null || roleNames.isEmpty()) {
+      return Collections.emptyList();
+    }
+
+    if (latestDropRoleTime.isBefore(latestGcRoleTime)) {
+      return roleNames;
+    }
+
+    List<RoleEntity> roleEntities = Lists.newArrayList();
+    int index = 0;
+    for (String role : roleNames) {
+      try {
+        RoleEntity roleEntity =
+            store.get(
+                NameIdentifierUtils.ofRole(metalake, role),
+                Entity.EntityType.ROLE,
+                RoleEntity.class);
+        if (roleEntity.id().equals(roleIds.get(index))) {
+          roleEntities.add(roleEntity);
+        }
+        index++;
 
-  private NameIdentifier ofGroup(String metalake, String group) {
-    return NameIdentifier.of(
-        metalake, Entity.SYSTEM_CATALOG_RESERVED_NAME, Entity.GROUP_SCHEMA_NAME, group);
+      } catch (IOException ioe) {
+        LOG.error(
+            "Checking roles {} failed in the metalake {} due to storage issues",
+            role,
+            metalake,
+            ioe);
+        throw new RuntimeException(ioe);
+      } catch (NoSuchEntityException nse) {
+        // ignore this entity
+      }
+    }
+    return roleEntities.stream().map(RoleEntity::name).collect(Collectors.toList());
   }
 }
```

Comment:
getValidXXX and removeInvalidXXX, it's the same thing, why do you implement twice?

Reply:
I will extract a common method for them.